apps/projects/routes.py

```diff
@@ -69,191 +34,216 @@ class ListUploadProject(MethodView):
-
-        check_user_agent()
-        schema = check_request_schema_validity(request.files, self.SCHEMA_UPLOAD)
+        document = validate_document(request.files, self.SCHEMA_UPLOAD)
 
         # validate codec
-        video_editor = get_video_editor()
-        file = schema['file']
-        file_stream = file.stream.read()
-        metadata = video_editor.get_meta(file_stream)
-        codec_name = metadata.get('codec_name')
-        if codec_name not in app.config.get('CODEC_SUPPORT'):
-            raise BadRequest("Codec: {} is not supported.".format(codec_name))
-        # generate file name
-        file_name = create_file_name(ext=file.filename.split('.')[1])
-        try:
-            # add record to database
-            doc = {
-                'filename': file_name,
-                'storage_id': None,
-                'metadata': metadata,
-                'create_time': datetime.utcnow(),
-                'mime_type': file.mimetype,
-                'version': 1,
-                'processing': False,
-                'parent': None,
-                'thumbnails': {},
-                'request_address': get_request_address(request.headers.environ),
-                'original_filename': file.filename,
-                'preview_thumbnail': None,
-                'url': None
+        file_stream = document['file'].stream.read()
+        metadata = get_video_editor().get_meta(file_stream)
+        if metadata.get('codec_name') not in app.config.get('CODEC_SUPPORT_VIDEO'):
+            raise BadRequest(f"Codec: '{metadata.get('codec_name')}' is not supported.")
+
+        # add record to database
+        project = {
+            'filename': create_file_name(ext=document['file'].filename.rsplit('.')[-1]),
+            'storage_id': None,
+            'metadata': metadata,
+            'create_time': datetime.utcnow(),
+            'mime_type': document['file'].mimetype,
+            'request_address': get_request_address(request.headers.environ),
+            'original_filename': document['file'].filename,
+            'version': 1,
+            'parent': None,
+            'processing': {
+                'video': False,
+                'thumbnail_preview': False,
+                'thumbnails_timeline': False
+            },
+            'thumbnails': {
+                'timeline': [],
+                'preview': None
             }
-            app.mongo.db.projects.insert_one(doc)
+        }
+        app.mongo.db.projects.insert_one(project)
 
-            # put file stream into storage
-            storage_id = app.fs.put(file_stream, file_name, doc['_id'], content_type=file.mimetype)
-            if not storage_id:
-                raise InternalServerError('Something went wrong when putting file into storage.')
+        # put file stream into storage
+        try:
+            storage_id = app.fs.put(
+                content=file_stream,
+                filename=project['filename'],
+                project_id=project['_id'],
```

Comment:
why is it needed here?

Reply:
and probably even if we need it, we can use `bson.ObjectId()` to get it and only insert that project once and not insert and update later.